app/plugin.go

```diff
@@ -259,13 +255,35 @@ func (a *App) SyncPlugins() *model.AppError {
 }
 
 func (a *App) ShutDownPlugins() {
-	pluginsEnvironment := a.GetPluginsEnvironment()
+	// Acquiring lock manually, as plugins might be disabled. See GetPluginsEnvironment.
+	a.Srv.PluginsLock.RLock()
+	pluginsEnvironment := a.Srv.PluginsEnvironment
+	a.Srv.PluginsLock.RUnlock()
+
 	if pluginsEnvironment == nil {
+		mlog.Warn("Plugins are already shutdown while trying to shut them down")
 		return
 	}
 
 	mlog.Info("Shutting down plugins")
 
+	availablePlugins, err := pluginsEnvironment.Available()
+	if err != nil {
+		mlog.Error("Failed to get available plugins while trying to shut them down. Clients will not get notified about shut down plugins.", mlog.Err(err))
+	}
+	for _, plugin := range availablePlugins {
+		pluginId := plugin.Manifest.Id
+		pluginEnabled := false
+		if state, ok := a.Config().PluginSettings.PluginStates[pluginId]; ok {
+			pluginEnabled = state.Enable
+		}
+
+		// Notifiy clients that the client has been shut down
+		if pluginEnabled && plugin.Manifest.HasClient() {
+			a.sendPluginDisableEvent(plugin.Manifest)
```

Comment:
hmm does this work? I think `a.Srv.Hubs` are already stopped by this point as looking through the shutdown flow in this [function](https://github.com/mattermost/mattermost-server/blob/master/app/server_app_adapters.go#L157)

We'll have to move `s.FakeApp().ShutDownPlugins()` to the top for this to work, no?